### ISSUE TITLE
Decoupling GraphicsDeviceManager from Game/GamePlatform/GameWindow

### ISSUE THREAD

**Issue by Jjagg** [MonoGame/MonoGame#5487]
The suggested changes here are probably a bit controversial and should go into a later MG 4.0 if accepted because of some breaking changes. This depends on #5486 and assumes the proposed changes there get merged.
  
If you look at GDM it could be completely decoupled from Game with relatively minor changes. The only thing really stopping this is the XNA API. The GDM constructor takes a Game as an argument so it can check that there's no GDM registered for the game yet and register itself in the games service container (1) and keep a reference for setting window handle in the PresentationParameters (2) and for handling display orientation (3). Let's try to handle these three things without requiring a Game!

1. I think we should let users manually register the GraphicsDeviceManager. I.e. instead of doing `graphics = new GraphicsDeviceManager(this)` in the Game1 ctor you'd do 
```
graphics = new GraphicsDeviceManager();
Services.AddService(typeof(IGraphicsDeviceManager), graphics);
Services.AddService(typeof(IGraphicsDeviceService), graphics);
```
and maybe shorthands for convenience:
```
graphics = new GraphicsDeviceManager();
this.RegisterGraphicsDeviceManager(graphics);
```
and
```
graphics = this.RegisterNewGraphicsDeviceManager();
```
Additionally this makes it more transparent to the user what actually happens.

2. XNA allows users to override the window handle in the PreparingPresentationParameters event, so it can be easily changed. Similar to what we do for other PresentationParameter stuff (like PreferredBackBufferWidth, PreferredBackBufferFormat, PreferredDepthStencilFormat...) I think it would be nice to have a WindowHandle property (public get and set) in GDM. That's more consistent and allows Game to set the initial window handle during initialization (only if the user has not set it yet , so they can do custom windowing stuff). I'd like to do some more research/thinking for this one though.

3. I haven't looked into MG code for mobile platforms a lot so I'm not entirely sure if the following is the best path, but I do have a suggestion. I think GDM should have a field for DisplayOrientation like it does for other stuff in presentation parameters (I already mentioned some in (2)). DisplayOrientation changes because of auto rotation on a device should change the value in GDM and apply changes. When DisplayMode is changed and `ApplyChanges` is called in code rotating the screen should be handled in `GamePlatform.OnPresentationChanged`.

The goal of these changes would be overall a lot more flexibility in using GraphicsDeviceManager. GDM would no longer depend on either Game, GameWindow or GamePlatform.

**Comment by Jjagg** [comment]
For anyone interested in this, I have an experimental branch over at https://github.com/Jjagg/MonoGame/tree/presentation2
I applied the changes I talked about here (and some more) and got things working for SDL. I'm pretty sure not all the changes in that branch will be/should be accepted in the core, but I'd like to try and get all platforms working on that branch and then we can see what can get merged in here. Changes I've made include:

- Got rid of the singletons in Game and GameWindow.
- Made GameWindow not depend on Game or GamePlatform.
- Made GraphicsDeviceManager totally unaware of Game (except for 1 method for compatibility, but I marked it as obsolete), GameWindow and GamePlatform.
- Changed BeginScreenDeviceChange/EndScreenDeviceChange API to OnPresentationChanged.

These changes probably broke every other platform for now :p 
Goals of these changes are to

1. Clean up code in Game/GameWindow/GamePlatform/GraphicsDeviceManager. Make the code more consistent and easier to understand and maintain.
2. Decouple classes were it makes sense, also makes maintaining easier.
3. Get rid of internal stuff so users can more easily extend MG without modifying the source code.
4. Change some quirky parts of the API to make them easier to understand for users.
5. Make embedding MonoGame easier.

but again, I'm not looking to get this all merged, it's just an experiment for now. It's very likely some of the API changes will cause issues for some platforms and it may end up worse than the current codebase in the end. But it's a fun experiment and we may get something useful out of it :)

**Comment by vpenades** [comment]
Along with decoupling these componets, I would propose to move all the code related to windows hosting to a separate DLL.

I would suggest a split similar to the one used by UrhoSharp

Urhosharp [Packages](https://www.nuget.org/packages?q=urhosharp)

Main Package:

- UrhoSharp

View Packages:

- UrhoSharp.WinForms
- UrhoSharp.WPF
- UrhoSharp.Cocoa
- UrhoSharp.Forms (Xamarin Forms, handling UWP, Android and iOS)
- UrhoSharp.SharpReality

Main package handles the engine and pure rendering, with DX in Windows and GL in all other platforms (no choice of graphics devices on windows desktop as far as I've seen)

View packages handle the creation and lifecycle of the hosting "window".

The advantage of this ecosystem is that it enforces the main package to provide everything a window hosting provider might require, as long as the hosting provider is able to give a GL or DX context.

For MonoGame, I would propose this split:

Main Package:

- Monogame.Framework

Client Graphics packages:

- MonoGame.ClientGraphics.DirectX
- MonoGame.ClientGraphics.OpenGL
- MonoGame.ClientGraphics.Vulkan (if someone wishes to write it!)

View Graphics packages:

- MonoGame.WinFormsGL
- MonoGame.WinFormsDX
- MonoGame.WPF (DX only, since WPF is DX under the hood)
- MonoGame.XamarinForms
- MonoGame.Android
- MonoGame.iOS
- MonoGame.UWP
- etc

If at some point we want to support VR, it will probably require custom views too, as it's the case of UrhoSharp.SharpReality.

For example:

- In Android's "VR Cardboard", the hosting Activity needs to inherit from VRCardboardActivity
- In Windows VR you need to go through the Mixed Reality SDK for window creation.

So my opinion is that the amount of technology permutations monogame is facing will stress more and more the current ecosystem, and some sort of split is bound to happen.
